crates/ra_hir/src/code_model.rs

```diff
@@ -1004,6 +1005,80 @@ impl ImplBlock {
     }
 }
 
+#[derive(Clone)]
+pub struct Type {
+    pub(crate) krate: CrateId,
+    pub(crate) ty: InEnvironment<Ty>,
+}
+
+impl Type {
+    pub fn fields(&self, db: &impl HirDatabase) -> Vec<(StructField, Type)> {
+        let mut res = Vec::new();
+        if let Ty::Apply(a_ty) = &self.ty.value {
+            match a_ty.ctor {
+                ty::TypeCtor::Adt(Adt::Struct(s)) => {
+                    for field in s.fields(db) {
+                        let ty = field.ty(db).subst(&a_ty.parameters);
+                        res.push((field, self.derived(ty)));
```

Comment:
Am I correct that we don't need to switch env here?

Ie, in the following example

```rust
struct Foo<T: Clone> {
  bar: Bar<T>
}

struct Bar<T>(T);
```

if the original env is `Foo`, when we do a `subst` we replace `Bar`'s `T` with our local `T` and the old env works?

Reply:
Yes. The environment is really a lexical thing, so if we're lexically within `Foo`, that doesn't change just because we're looking at another type.